src/main/java/com/lginterview/providers/CourseProviderImpl.java

```diff
@@ -0,0 +1,139 @@
+            return courseActions.get(courseRequest.getType()).apply(courseRequest);
+        else
+            throw new IllegalArgumentException("Invalid Course Request type");
+    }
+
+    private Long createCourse(CourseRequest courseRequest) throws CourseDataRetrievalException {
+        Course newCourse = courseRepository.findOne(courseRequest.getCourseId());
+        if (newCourse != null)
+            return newCourse.getId();
+
+        User courseUser = userRepository.findOne(courseRequest.getUserId());
+        if (courseUser == null)
+            throw new CourseDataRetrievalException("User not found");
+
+        newCourse = new Course();
+        newCourse.setStartDate(courseRequest.getTimestamp());
+        newCourse.setCourseUser(courseUser);
+        return courseRepository.save(newCourse).getId();
+    }
+
+    private Long updateCourse(CourseRequest courseRequest) throws CourseDataRetrievalException {
+
+        Course courseToUpdate = courseRepository.findOne(courseRequest.getCourseId());
+        if (courseToUpdate == null)
+            throw new CourseDataRetrievalException("Course Not Found");
+
+        if (courseRequest.getScore() != null) {
+            int scoreValue = courseRequest.getScore();
+            courseToUpdate.getScores().add(scoreValue);
+            courseRepository.save(courseToUpdate);
+        }
+        return courseToUpdate.getId();
+    }
+
+    private Long closeCourse(CourseRequest courseRequest) throws CourseDataRetrievalException {
+        Course courseToClose = courseRepository.findOne(courseRequest.getCourseId());
+        if (courseToClose == null)
+            throw new CourseDataRetrievalException("Course Not Found");
+
+        courseToClose.setEndDate(courseRequest.getTimestamp());
+        courseRepository.save(courseToClose);
+        return courseToClose.getId();
+    }
+
+    @Override
+    @Transactional(readOnly = true)
+    public CourseStatus getCourseStatus(long userId, long courseId) throws CourseDataRetrievalException {
+
+        Course courseData = getCourse(userId, courseId);
+
+        int bestScore = 0;
+        int averageScore = 0;
+
+        if ((courseData.getScores() != null)) {
+            bestScore = courseData.getScores()
+                    .stream()
+                    .mapToInt(score -> score)
+                    .max()
+                    .orElse(0);
+
+            averageScore = (int) courseData.getScores()
+                    .stream()
+                    .mapToInt(score -> score)
+                    .average()
+                    .orElse(0.0);
+        }
+
+        int courseTime = 0;
+        if (courseData.getEndDate() != null)
+            courseTime = (int) (courseData.getEndDate().toEpochSecond() - courseData.getStartDate().toEpochSecond());
+
+        ZonedDateTime utcStartTime=getUTCDateTime(courseData.getStartDate());
+        ZonedDateTime utcEndTime=getUTCDateTime(courseData.getEndDate());
```

Comment:
[MS][Minor] You should not need to do this, because the whole application/business logic should exclusively operate on UTC time, to not have to think about conversion to UTC all the time.

Reply:
I wrote a comment about this earlier